db/compaction/compaction.cc

```diff
@@ -415,6 +415,52 @@ void Compaction::MarkFilesBeingCompacted(bool mark_as_compacted) {
   }
 }
 
+Slice Compaction::StartLevelSmallestInternalKey() const {
+  // lazy initialize, return if it's already calculated
+  if (!start_level_smallest_internal_.empty()) {
+    return start_level_smallest_internal_;
+  }
+
+  assert(inputs_[0].level == start_level());
+  assert(!inputs_[0].files.empty());
+  // non-L0 inputs are sorted
+  start_level_smallest_internal_ = inputs_[0].files.front()->smallest.Encode();
```

Comment:
I don't have to do it now, but I think we might consider to allow writing to the gap part too, if one day we would allow data from the last level to be moved to a higher level. Otherwise, some range of the last level might never get a chance to be moved up. This is not something can be solved in this function, so maybe add a TODO.

Reply:
Added a TODO for moving data from the last level to up level. For time-based tiering, it should never happen. But for range base tiering, it could happen.